pytorch_lightning/trainer/training_loop.py

```diff
@@ -188,10 +189,10 @@ def reset_train_val_dataloaders(self, model) -> None:
         The val dataloader must be initialized before training loop starts, as the training loop
         inspects the val dataloader to determine whether to run the evaluation loop.
         """
-        if self.trainer.train_dataloader is None:
+        if self.trainer.train_dataloader is None or self.trainer.state.status == TrainerStatus.RUNNING:
             self.trainer.reset_train_dataloader(model)
 
-        if self.trainer.val_dataloaders is None:
+        if self.trainer.val_dataloaders is None or self.trainer.tuning:
```

Comment:
if this fix works, it tells me that the problem is somewhere else

shouldn't have to check this here.

Reply:
Sorry, I thought it is "TunerStatus". Yes, you are right, it should always RUNNING. In fact that, I had one solution that adding `or self.trainer.tuning` for both train and val dataloaders. And switch 

self.trainer.training = True       
self.trainer._run(*args, **kwargs) 
self.trainer.tuning = True

to

self.trainer.training = True       
self.trainer.tuning = True
self.trainer._run(*args, **kwargs) 

in

https://github.com/PyTorchLightning/pytorch-lightning/blob/20f63377f81f4771d3f128f979b3a0f9b8d219a7/pytorch_lightning/tuner/tuning.py#L60-L65
But it breaks one existing test in test_model_state
https://github.com/PyTorchLightning/pytorch-lightning/blob/20f63377f81f4771d3f128f979b3a0f9b8d219a7/tests/trainer/test_states.py#L50
Do you think we should keep this direction to fix this problem and work on solving the test issue? Because we need to reset the dataloder for each `tuner._run()` -> `trainer._run()`. Because the dataloader is attached only once before the tuning, hence we need to reset the dataloder even it is not none when it is in tuning status.

Reply:
> Do you think we should keep this direction to fix this problem and work on solving the test issue? 

No, that fix would also not be correct. Due to how tuning works, we need to set the trainer stage to training before the tuner calls `_run` internally.

If you can add the failing test that covers the bug, I can have a look and try to resolve it.

Reply:
Yes, I understand that. Because the model stage can only be either Runningtage.TUNING or RunningStage.TRAINING. 
Hence, if I move `tuning = True` before `_run`, it results in the property of `trainer.training` becomes False.  
I think we need somehow let logic inside `trainer._run` know it is called from tunner and reset the dataloader's. Or we need to find a way to reset the dataloader before calling `_run` in Tuner. What do you think?

Reply:
I am thinking, what about just simply call reset_train_dataloder in `tuner._run`? 
```python
   def _run(self, *args: Any, **kwargs: Any) -> None:
        """`_run` wrapper to set the proper state during tuning, as this can be called multiple times"""
        self.trainer.state.status = TrainerStatus.RUNNING  # last `_run` call might have set it to `FINISHED`
        self.trainer.training = True
        self.trainer.reset_train_dataloader(model)
        self.trainer._run(*args, **kwargs)
        self.trainer.tuning = True
```

Reply:
Can we check if the `trainer.train_loop.global_step == 0` in `reset_train_val_dataloaders()`

Reply:
Hi @carmocca , what do you think if we simply reset `trainer.train_dataloader` to None at each interaction step when searching for batch size? 
```python
def _run_power_scaling(
    trainer: 'pl.Trainer', model: 'pl.LightningModule', new_size: int, batch_arg_name: str, max_trials: int
) -> int:
    """ Batch scaling mode where the size is doubled at each iteration until an OOM error is encountered. """
    for _ in range(max_trials):
        garbage_collection_cuda()
        trainer.train_loop.global_step = 0  # reset after each try
        try:
            #!!!! Set trainer train_dataloader to None in order to reset it
            trainer.train_dataloader = None
            # Try fit
            trainer.tuner._run(model)
            # Double in size
            new_size, changed = _adjust_batch_size(trainer, batch_arg_name, factor=2.0, desc='succeeded')
        except RuntimeError as exception:
            # Only these errors should trigger an adjustment
            if is_oom_error(exception):
                # If we fail in power mode, half the size and return
                garbage_collection_cuda()
                new_size, _ = _adjust_batch_size(trainer, batch_arg_name, factor=0.5, desc='failed')
                break
            else:
                raise  # some other error not memory related

        if not changed:
            break
    return new_size
```
Same for `_run_binsearch_scaling`.

Reply:
Yes, that seems like a better option to me.

```python
if changed:
    # set dataloaders to None so they are reset
else:
    break

Reply:
@carmocca shouldn't it be:

```
for _ in range(max_trials):
    garbage_collection_cuda()
    trainer.train_loop.global_step = 0  # reset after each try
    try:
        # Try fit
        trainer.tuner._run(model)
        # Double in size
        new_size, changed = _adjust_batch_size(trainer, batch_arg_name, factor=2.0, desc='succeeded')
    except RuntimeError as exception:
        # Only these errors should trigger an adjustment
        if is_oom_error(exception):
            # If we fail in power mode, half the size and return
            garbage_collection_cuda()
            new_size, _ = _adjust_batch_size(trainer, batch_arg_name, factor=0.5, desc='failed')
            break
        else:
            raise  # some other error not memory related

    if not changed:
        break

if changed:
    trainer.train_dataloader = None

return new_size
```